### PR TITLE
Improve Conclusions Help/Documentation #452

### PR DESCRIPTION
**Pull request by kopal960**
solved : #452 
- In edit conclusions modal
   - I have edited the help text of all scan based fields and added an icon for them
     Screenshot 2020-12-23 200003

   - If the scan does not have any value for a particular field , the following text will be displayed. This is a result of the changes in assets/app/js/controllers/conclusionDialog.js
     Screenshot 2020-12-23 200116
Is there anything else that I can add or change?

### CODE REVIEW

File: assets/app/js/controllers/conclusionDialog.js [hunk 1]
```diff
@@ -204,7 +204,12 @@ class ConclusionDialog extends Controller {
           data: $.unique(license_expressions),
           multiple: true,
           placeholder: 'Enter Expression',
-          tags: true
+          tags: true,
+          language: {
+            'noResults': function () {
+              return (`SCWB did not find any license expressions in the scan that applies to ${conclusion.path}`);
```
Comment (steven-esser) [inline assets/app/js/controllers/conclusionDialog.js]: In some cases, this `conclusion.path` value is very long and is awkward to show in a small menu item. 

Instead of `${conclusion.path}`, we should have "the current path".

File: assets/app/js/controllers/conclusionDialog.js [hunk 2]
```diff
@@ -221,7 +226,12 @@ class ConclusionDialog extends Controller {
           data: $.unique(copyright_statements),
           multiple: true,
           placeholder: 'Enter copyright',
-          tags: true
+          tags: true,
+          language: {
+            'noResults': function () {
+              return (`SCWB did not find any copyright statements in the scan pertaining to ${conclusion.path}`);
```
Comment (steven-esser) [inline assets/app/js/controllers/conclusionDialog.js]: In some cases, this `conclusion.path` value is very long and is awkward to show in a small menu item. 

Instead of `${conclusion.path}`, we should have "the current path".

File: assets/app/js/controllers/conclusionDialog.js [hunk 3]
```diff
@@ -237,7 +247,12 @@ class ConclusionDialog extends Controller {
           multiple: true,
           maximumSelectionLength: 1,
           placeholder: 'Enter owner',
-          tags: true
+          tags: true,
+          language: {
+            'noResults': function () {
+              return (`SCWB did not find any owners in the scan for ${conclusion.path}`);
```
Comment (steven-esser) [inline assets/app/js/controllers/conclusionDialog.js]: In some cases, this `conclusion.path` value is very long and is awkward to show in a small menu item. 

Instead of `${conclusion.path}`, we should have "the current path".

File: assets/app/js/controllers/conclusionDialog.js [hunk 4]
```diff
@@ -254,7 +269,12 @@ class ConclusionDialog extends Controller {
           multiple: true,
           maximumSelectionLength: 1,
           placeholder: 'Enter language',
-          tags: true
+          tags: true,
+          language: {
+            'noResults': function () {
+              return (`SCWB did not find any primary programming languages in the scan associated with ${conclusion.path}`);
```
Comment (steven-esser) [inline assets/app/js/controllers/conclusionDialog.js]: In some cases, this `conclusion.path` value is very long and is awkward to show in a small menu item. 

Instead of `${conclusion.path}`, we should have "the current path".

File: assets/app/js/controllers/conclusionDialog.js [hunk 5]
```diff
@@ -283,7 +303,12 @@ class ConclusionDialog extends Controller {
           multiple: true,
           maximumSelectionLength: 1,
           placeholder: 'Enter Homepage URL',
-          tags: true
+          tags: true,
+          language: {
+            'noResults': function () {
+              return (`SCWB did not find any Homepage URL in the scan for ${conclusion.path}`);
```
Comment (steven-esser) [inline assets/app/js/controllers/conclusionDialog.js]: In some cases, this `conclusion.path` value is very long and is awkward to show in a small menu item. 

Instead of `${conclusion.path}`, we should have "the current path".

File: assets/app/js/controllers/conclusionDialog.js [hunk 6]
```diff
@@ -303,7 +328,12 @@ class ConclusionDialog extends Controller {
           multiple: true,
           maximumSelectionLength: 1,
           placeholder: 'Enter Download URL',
-          tags: true
+          tags: true,
+          language: {
+            'noResults': function () {
+              return (`SCWB did not find any download URL in the scan for obtaining ${conclusion.path}`);
```
Comment (steven-esser) [inline assets/app/js/controllers/conclusionDialog.js]: In some cases, this `conclusion.path` value is very long and is awkward to show in a small menu item. 

Instead of `${conclusion.path}`, we should have "the current path".

File: assets/app/js/controllers/conclusionDialog.js [hunk 7]
```diff
@@ -325,7 +355,12 @@ class ConclusionDialog extends Controller {
           multiple: true,
           maximumSelectionLength: 1,
           placeholder: 'Enter License URL',
-          tags: true
+          tags: true,
+          language: {
+            'noResults': function () {
+              return (`SCWB did not find any License url in the scan for ${conclusion.path}`);
```
Comment (steven-esser) [inline assets/app/js/controllers/conclusionDialog.js]: In some cases, this `conclusion.path` value is very long and is awkward to show in a small menu item. 

Instead of `${conclusion.path}`, we should have "the current path".

File: assets/app/js/controllers/conclusionDialog.js [hunk 8]
```diff
@@ -341,7 +376,12 @@ class ConclusionDialog extends Controller {
           multiple: true,
           maximumSelectionLength: 1,
           placeholder: 'Enter Notice URL',
-          tags: true
+          tags: true,
+          language: {
+            'noResults': function () {
+              return (`SCWB did not find any Notice URL in the scan for ${conclusion.path}`);
```
Comment (steven-esser) [inline assets/app/js/controllers/conclusionDialog.js]: In some cases, this `conclusion.path` value is very long and is awkward to show in a small menu item. 

Instead of `${conclusion.path}`, we should have "the current path".

File: assets/thirdparty/fonts/lato.LICENSE [hunk 9]
```diff
@@ -1,94 +1,94 @@
...
+DISCLAIMER
+THE FONT SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO ANY WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT
+OF COPYRIGHT, PATENT, TRADEMARK, OR OTHER RIGHT. IN NO EVENT SHALL THE
+COPYRIGHT HOLDER BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+INCLUDING ANY GENERAL, SPECIAL, INDIRECT, INCIDENTAL, OR CONSEQUENTIAL
+DAMAGES, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF THE USE OR INABILITY TO USE THE FONT SOFTWARE OR FROM
+OTHER DEALINGS IN THE FONT SOFTWARE.
```
Comment (steven-esser) [inline assets/thirdparty/fonts/lato.LICENSE]: Why is this file modified?

File: assets/thirdparty/fonts/raleway.LICENSE [hunk 10]
```diff
@@ -1,94 +1,94 @@
...
+DISCLAIMER
+THE FONT SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO ANY WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT
+OF COPYRIGHT, PATENT, TRADEMARK, OR OTHER RIGHT. IN NO EVENT SHALL THE
+COPYRIGHT HOLDER BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+INCLUDING ANY GENERAL, SPECIAL, INDIRECT, INCIDENTAL, OR CONSEQUENTIAL
+DAMAGES, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF THE USE OR INABILITY TO USE THE FONT SOFTWARE OR FROM
+OTHER DEALINGS IN THE FONT SOFTWARE.
```
Comment (steven-esser) [inline assets/thirdparty/fonts/raleway.LICENSE]: Why is this file modified?

File: index.html [hunk 11]
```diff
@@ -394,15 +394,15 @@ <h3 class="modal-title">Modal title</h3>
                             <input type="text" class="form-control" id="conclusion-version" name="conclusion-version" placeholder="Enter version">
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-license-expression" data-toggle="tooltip" data-placement="top" title="The license expression that applies to a conclusion.">License Expression</label>
+                            <label for="conclusion-license-expression" data-toggle="tooltip" data-placement="top" title="The license expression that applies to a conclusion.SCWB will present possible values from the scan.">License Expression   <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There needs to be a space before `SCWB`

File: index.html [hunk 12]
```diff
@@ -394,15 +394,15 @@ <h3 class="modal-title">Modal title</h3>
...
                             <select data-hide-on-select="true" class="form-control" id="conclusion-owner" name="conclusion-owner" type="text"></select>
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-copyright" data-toggle="tooltip" data-placement="top" title="The copyright statement(s) that pertain to this conclusion, as contained in the source or as specified in an associated file.">Copyright</label>
+                            <label for="conclusion-copyright" data-toggle="tooltip" data-placement="top" title="The copyright statement(s) that pertain to this conclusion, as contained in the source or as specified in an associated file.SCWB will present possible values from the scan.">Copyright    <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There needs to be a space before `SCWB`

File: index.html [hunk 13]
```diff
@@ -444,23 +444,23 @@ <h3 class="modal-title">Modal title</h3>
...
+                            <label for="conclusion-language" data-toggle="tooltip" data-placement="top" title="The primary programming language associated with the conclusion.SCWB will present possible values from the scan.">Programming Language   <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
                             <select data-hide-on-select="true" class="form-control" id="conclusion-language" name="conclusion-language" type="text"></select>
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-homepage-url" data-toggle="tooltip" data-placement="top" title="Homepage URL for the conclusion.">Homepage URL</label>
+                            <label for="conclusion-homepage-url" data-toggle="tooltip" data-placement="top" title="Homepage URL for the conclusion.SCWB will present possible values from the scan.">Homepage URL   <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There needs to be a space before `SCWB`

File: index.html [hunk 14]
```diff
@@ -444,23 +444,23 @@ <h3 class="modal-title">Modal title</h3>
...
                             <select data-hide-on-select="true" class="form-control" id="conclusion-homepage-url" name="conclusion-homepage-url" type="text"></select>
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-download-url" data-toggle="tooltip" data-placement="top" title="The download URL for obtaining the package.">Download URL</label>
+                            <label for="conclusion-download-url" data-toggle="tooltip" data-placement="top" title="The download URL for obtaining the package.SCWB will present possible values from the scan.">Download URL    <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There needs to be a space before `SCWB`

File: index.html [hunk 15]
```diff
@@ -444,23 +444,23 @@ <h3 class="modal-title">Modal title</h3>
                             <input type="text" class="form-control" id="conclusion-purpose" name="conclusion-purpose" placeholder="Enter purpose">
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-language" data-toggle="tooltip" data-placement="top" title="The primary programming language associated with the conclusion.">Programming Language</label>
+                            <label for="conclusion-language" data-toggle="tooltip" data-placement="top" title="The primary programming language associated with the conclusion.SCWB will present possible values from the scan.">Programming Language   <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There needs to be a space before `SCWB`

File: index.html [hunk 16]
```diff
@@ -394,15 +394,15 @@ <h3 class="modal-title">Modal title</h3>
...
                             <select data-hide-on-select="true" class="form-control " id="conclusion-license-expression" name="conclusion-license-expression" type="text"></select>
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-owner" data-toggle="tooltip" data-placement="top" title="The unique user-maintained name of the author, custodian, or provider of one or more software objects (licenses, products).">Owner</label>
+                            <label for="conclusion-owner" data-toggle="tooltip" data-placement="top" title="The unique user-maintained name of the author, custodian, or provider of one or more software objects (licenses, products).SCWB will present possible values from the scan.">Owner  <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There needs to be a space before `SCWB`

File: index.html [hunk 17]
```diff
@@ -444,23 +444,23 @@ <h3 class="modal-title">Modal title</h3>
...
+                            <label for="conclusion-download-url" data-toggle="tooltip" data-placement="top" title="The download URL for obtaining the package.SCWB will present possible values from the scan.">Download URL    <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
                             <select data-hide-on-select="true" class="form-control" id="conclusion-download-url" name="conclusion-download-url" type="text"></select>
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-license-url" data-toggle="tooltip" data-placement="top" title="Where available, the URL to the license text.">License URL</label>
+                            <label for="conclusion-license-url" data-toggle="tooltip" data-placement="top" title="Where available, the URL to the license text.SCWB will present possible values from the scan.">License URL    <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There needs to be a space before `SCWB`

File: index.html [hunk 18]
```diff
@@ -444,23 +444,23 @@ <h3 class="modal-title">Modal title</h3>
...
                             <select data-hide-on-select="true" class="form-control" id="conclusion-license-url" name="conclusion-license-url" type="text"></select>
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-notice-url" data-toggle="tooltip" data-placement="top" title="A URL that contains notice text for a conclusion.">Notice URL</label>
+                            <label for="conclusion-notice-url" data-toggle="tooltip" data-placement="top" title="A URL that contains notice text for a conclusion.SCWB will present possible values from the scan.">Notice URL  <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There needs to be a space before `SCWB`

**Review by steven-esser — CHANGES_REQUESTED** [review]
@kopal960 Thanks for this. 

See PR comments.

File: index.html (hunk 18 above)
Comment (kopal960) [reply index.html]: Okay , I'll change that.

File: assets/thirdparty/fonts/raleway.LICENSE (hunk 10 above)
Comment (kopal960) [reply assets/thirdparty/fonts/raleway.LICENSE]: This was a result of changing the line endings ,I'll revert back this change.

File: assets/thirdparty/fonts/lato.LICENSE (hunk 9 above)
Comment (kopal960) [reply assets/thirdparty/fonts/lato.LICENSE]: This was a result of changing the line endings ,I'll revert back this change.

File: index.html [hunk 19]
```diff
@@ -394,15 +394,15 @@ <h3 class="modal-title">Modal title</h3>
                             <input type="text" class="form-control" id="conclusion-version" name="conclusion-version" placeholder="Enter version">
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-license-expression" data-toggle="tooltip" data-placement="top" title="The license expression that applies to a conclusion.">License Expression</label>
+                            <label for="conclusion-license-expression" data-toggle="tooltip" data-placement="top" title="The license expression that applies to a conclusion. SCWB will present possible values from the scan.">License Expression   <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There seems to be some extraneous whitespace after the label's text?

File: index.html [hunk 20]
```diff
@@ -394,15 +394,15 @@ <h3 class="modal-title">Modal title</h3>
...
                             <select data-hide-on-select="true" class="form-control " id="conclusion-license-expression" name="conclusion-license-expression" type="text"></select>
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-owner" data-toggle="tooltip" data-placement="top" title="The unique user-maintained name of the author, custodian, or provider of one or more software objects (licenses, products).">Owner</label>
+                            <label for="conclusion-owner" data-toggle="tooltip" data-placement="top" title="The unique user-maintained name of the author, custodian, or provider of one or more software objects (licenses, products). SCWB will present possible values from the scan.">Owner  <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There seems to be some extraneous whitespace after the label's text?

File: index.html [hunk 21]
```diff
@@ -394,15 +394,15 @@ <h3 class="modal-title">Modal title</h3>
...
                             <select data-hide-on-select="true" class="form-control" id="conclusion-owner" name="conclusion-owner" type="text"></select>
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-copyright" data-toggle="tooltip" data-placement="top" title="The copyright statement(s) that pertain to this conclusion, as contained in the source or as specified in an associated file.">Copyright</label>
+                            <label for="conclusion-copyright" data-toggle="tooltip" data-placement="top" title="The copyright statement(s) that pertain to this conclusion, as contained in the source or as specified in an associated file. SCWB will present possible values from the scan.">Copyright    <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There seems to be some extraneous whitespace after the label's text?

File: index.html [hunk 22]
```diff
@@ -444,23 +444,23 @@ <h3 class="modal-title">Modal title</h3>
                             <input type="text" class="form-control" id="conclusion-purpose" name="conclusion-purpose" placeholder="Enter purpose">
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-language" data-toggle="tooltip" data-placement="top" title="The primary programming language associated with the conclusion.">Programming Language</label>
+                            <label for="conclusion-language" data-toggle="tooltip" data-placement="top" title="The primary programming language associated with the conclusion. SCWB will present possible values from the scan.">Programming Language   <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There seems to be some extraneous whitespace after the label's text?

File: index.html [hunk 23]
```diff
@@ -444,23 +444,23 @@ <h3 class="modal-title">Modal title</h3>
...
                             <select data-hide-on-select="true" class="form-control" id="conclusion-language" name="conclusion-language" type="text"></select>
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-homepage-url" data-toggle="tooltip" data-placement="top" title="Homepage URL for the conclusion.">Homepage URL</label>
+                            <label for="conclusion-homepage-url" data-toggle="tooltip" data-placement="top" title="Homepage URL for the conclusion. SCWB will present possible values from the scan.">Homepage URL   <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There seems to be some extraneous whitespace after the label's text?

File: index.html [hunk 24]
```diff
@@ -444,23 +444,23 @@ <h3 class="modal-title">Modal title</h3>
...
                             <select data-hide-on-select="true" class="form-control" id="conclusion-homepage-url" name="conclusion-homepage-url" type="text"></select>
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-download-url" data-toggle="tooltip" data-placement="top" title="The download URL for obtaining the package.">Download URL</label>
+                            <label for="conclusion-download-url" data-toggle="tooltip" data-placement="top" title="The download URL for obtaining the package. SCWB will present possible values from the scan.">Download URL    <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There seems to be some extraneous whitespace after the label's text?

File: index.html [hunk 25]
```diff
@@ -444,23 +444,23 @@ <h3 class="modal-title">Modal title</h3>
...
                             <select data-hide-on-select="true" class="form-control" id="conclusion-download-url" name="conclusion-download-url" type="text"></select>
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-license-url" data-toggle="tooltip" data-placement="top" title="Where available, the URL to the license text.">License URL</label>
+                            <label for="conclusion-license-url" data-toggle="tooltip" data-placement="top" title="Where available, the URL to the license text. SCWB will present possible values from the scan.">License URL    <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There seems to be some extraneous whitespace after the label's text?

File: index.html [hunk 26]
```diff
@@ -444,23 +444,23 @@ <h3 class="modal-title">Modal title</h3>
...
                             <select data-hide-on-select="true" class="form-control" id="conclusion-license-url" name="conclusion-license-url" type="text"></select>
                         </fieldset>
                         <fieldset class="form-group">
-                            <label for="conclusion-notice-url" data-toggle="tooltip" data-placement="top" title="A URL that contains notice text for a conclusion.">Notice URL</label>
+                            <label for="conclusion-notice-url" data-toggle="tooltip" data-placement="top" title="A URL that contains notice text for a conclusion. SCWB will present possible values from the scan.">Notice URL  <i class="fa fa-check-circle" aria-hidden="true" title="scan-based"></i></label>
```
Comment (steven-esser) [inline index.html]: There seems to be some extraneous whitespace after the label's text?

**Review by steven-esser — CHANGES_REQUESTED** [review]
@kopal960 This is almost there, just some minor formatting/whitespace problems.

**Comment by kopal960** [comment]
> @kopal960 This is almost there, just some minor formatting/whitespace problems.

Okay ,so I'll reduce those extraneous spaces to one character space?

File: index.html (hunk 26 above)
Comment (kopal960) [reply index.html]: > There seems to be some extraneous whitespace after the label's text?

Okay ,so I'll reduce those extraneous spaces to one character space?

Comment (steven-esser) [reply index.html]: We want to have no trailing white space at all.

**Review by steven-esser — CHANGES_REQUESTED** [review]
@kopal960 The code content is now perfect, thanks for the contribution!

There is one final request to make before I merge this PR. Can you please rebase your work into a single commit? Right now, you seem to have many "intermediate" commits, like `Indentation fixed` etc. Having a single commit message that gives a bit of detail on your change is better than a flood of tiny commits. 

Also, see: https://chris.beams.io/posts/git-commit/ for more information on how to write a good git commit message.

**Comment by kopal960** [comment]
@majurg  Will do that .Thanks for the resource.

**Review by steven-esser — APPROVED** [review]
